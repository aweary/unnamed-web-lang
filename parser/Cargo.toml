[package]
name = "parser"
version = "0.1.0"
authors = ["Brandon Dail <brandondail@fb.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
graphql-parser = "0.2.3"
salsa = "0.14.1"
lexer = { path = "../lexer" }
syntax = { path = "../syntax" }
diagnostics = { path = "../diagnostics" }
source = { path = "../source" }
insta = { version = "0.16", features = ["glob"] }
log = "0.4.8"