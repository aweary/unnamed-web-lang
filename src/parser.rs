use ast::*;
use lexer::Lexer;
use token::*;

pub struct Parser {
    pub lexer: Lexer,
    // state: ParserState,
}

impl Parser {
    pub fn module(source: String) -> Self {
        let lexer = Lexer::new(source);
        // let state = ParserState::new();
        Parser { lexer }
    }

    fn peek(&mut self) -> Option<&Token> {
        self.lexer.peek_token()
    }

    pub fn skip_until<F>(&mut self, predicate: F)
    where
        F: Fn(Token) -> bool,
    {
        loop {
            match self.lexer.next_token() {
                Some(token) => {
                    if predicate(token) {
                        return;
                    }
                }
                None => return,
            }
        }
    }

    pub fn skip_assert(&mut self, token: Token) {
        assert_eq!(Some(token), self.lexer.next_token())
    }

    fn read_param_list(&mut self) -> Option<Vec<Param>> {
        match self.lexer.next_token() {
            // If the next token is an opening curly bracket, there
            // are no arguments and we assume this component takes no input
            Some(Token::LCurlyBracket) => None,
            // If the next token is an opening paren, there is
            // an argument list.
            Some(Token::LParen) => {
                let mut params: Vec<Param> = vec![];
                loop {
                    let name = self.read_ident();
                    // TODO this should be an unexpected token error
                    assert_eq!(Some(Token::Colon), self.lexer.next_token());
                    let kind = self.read_ident();
                    params.push(Param::new(name, kind));
                    match self.lexer.next_token() {
                        Some(Token::RParen) => {
                            break;
                            // End of the param list
                        }
                        Some(Token::Comma) => {
                            continue;
                            // There's another paramter
                        }
                        // Anything else is an unexpected token error
                        _ => unreachable!(),
                    }
                }
                Some(params)
            }
            _ => unreachable!("Unexpected token"),
        }
        // Next token should be the opening paren for the argument list
    }

    fn read_ident(&mut self) -> String {
        match self.lexer.next_token() {
            Some(Token::Ident(name)) => name,
            unknown @ _ => unreachable!("Expected identifier, found: {:?}", unknown),
        }
    }

    // Assumes that the current token is Reserved(Component)
    fn read_component_declaration(&mut self) -> Node {
        // Next token should be the identifier for the component definition
        let name = self.read_ident();
        // Read list of all params
        let params = self.read_param_list();
        // Everything from here until the closing bracket is part of the
        // component definition
        let body = self.read_component_body();
        let decl = ComponentDefinition::new(name, params, body);
        Node::ComponentDefinition(decl)
    }

    // A component is composed of a set of expressions/statements.
    fn read_component_body(&mut self) -> Vec<Expression> {
        let mut exprs = vec![];
        loop {
            match self.lexer.next_token() {
                // Statements starting with '<' are assumed to be
                // the beginning of JSX statements.
                Some(Token::LessThan) => {
                    let expr = Expression::JSXElement(self.read_jsx_element());
                    exprs.push(expr);
                    return exprs;
                }
                _ => {
                    continue;
                }
            }
        }
    }

    /**
     * read_expr
     * Reads any expresison, which is something that can evaluate to some value
     */
    fn read_expr(&mut self) -> Expression {
        self.skip_until(|token| token == Token::RCurlyBracket);
        Expression::NumericLiteral(42)
    }

    fn read_jsx_attributes(&mut self) -> Option<Vec<JSXAttribute>> {
        let mut attrs: Vec<JSXAttribute> = vec![];
        loop {
            match self.peek() {
                Some(&Token::Ident(_)) => {
                    let attr = self.read_jsx_attribute();
                    attrs.push(attr);
                }
                Some(&Token::ForwardSlash) | Some(&Token::GreaterThan) => {
                    break;
                }
                _ => {
                    unreachable!("Unknown token: {:?}", self.peek());
                }
            }
        }
        println!("read_jsx_attributes: {:?}", attrs);
        if attrs.is_empty() {
            None
        } else {
            Some(attrs)
        }
    }

    fn is_jsx_element_self_closing(&mut self) -> bool {
        match self.peek() {
            Some(&Token::GreaterThan) => {
                self.skip_assert(Token::GreaterThan);
                false
            }
            Some(&Token::ForwardSlash) => {
                self.skip_assert(Token::ForwardSlash);
                self.skip_assert(Token::GreaterThan);
                true
            }
            _ => {
                unreachable!();
            }
        }
    }

    fn read_jsx_element(&mut self) -> JSXElement {
        let tag_name = self.read_ident();
        let attributes = self.read_jsx_attributes();
        let self_closing = self.is_jsx_element_self_closing();
        let opening_element = JSXOpeningElement::new(tag_name.clone(), attributes, self_closing);
        println!("opening_element {:?}", opening_element);
        let (children, closing_element) = if self_closing {
            (None, None)
        } else {
            let children = self.read_jsx_children();
            let closing_element = self.read_jsx_closing_element();
            println!("closing_element {:?}", closing_element);
            println!("children {:?}", children);
            (Some(children), Some(closing_element))
        };
        let element = JSXElement::new(opening_element, closing_element, children);
        println!("element: {:#?}", element);
        element
    }

    fn read_jsx_closing_element(&mut self) -> JSXClosingElement {
        // Closing tag
        // self.skip_assert(Token::LessThan);
        // self.skip_assert(Token::ForwardSlash);
        let closing_tag_name = self.read_ident();
        // TODO parsing error
        // assert_eq!(tag_name, closing_tag_name);
        self.skip_assert(Token::GreaterThan);
        // TODO actually parse the closing tag
        JSXClosingElement::new(closing_tag_name)
    }

    fn read_jsx_text(&mut self) -> Box<JSXChildren> {
        let mut text = String::new();
        loop {
            match self.peek() {
                Some(&Token::Ident(_)) => {
                    let word = self.read_ident();
                    // TODO respect whitespace when lexing instead of this hack
                    text = text + " " + &word;
                }
                Some(&Token::LessThan) | Some(&Token::LCurlyBracket) => {
                    break;
                }
                _ => {
                    unreachable!();
                }
            }
        }
        println!("read_jsx_text: {:?}", text);
        Box::new(JSXChildren::JSXText(text))
    }

    fn read_jsx_children(&mut self) -> Vec<Box<JSXChildren>> {
        let mut children: Vec<Box<JSXChildren>> = vec![];
        loop {
            match self.peek() {
                // JSXText
                Some(&Token::Ident(_)) => {
                    children.push(self.read_jsx_text());
                },
                Some(&Token::LessThan) => {
                    // A '<' token could be the start of another element
                    // or the closing element for the enclosing element
                    self.skip_assert(Token::LessThan);
                    match self.peek() {
                        Some(&Token::Ident(_)) => {
                            // An identifier means the start of a new element
                            let element = self.read_jsx_element();
                            children.push(Box::new(JSXChildren::JSXElement(element)));
                        }
                        Some(&Token::ForwardSlash) => {
                            self.skip_assert(Token::ForwardSlash);
                            return children;
                            // A '/' indicates a closing element
                        }
                        _ => {
                            unreachable!();
                        }
                    }
                }
                _ => break,
            }
        }
        children
        // let child = match token {
        //     // TODO JSX parsers respect the whitespace inside tags. Need to output tokens
        //     Some(Token::Ident(word)) => {
        //         JSXChildren::JSXText(word)
        //     }
        //     Some(Token::LessThan) => {
        //         let element = self.read_jsx_element();
        //         JSXChildren::JSXElement(element)
        //     }
        //     // TODO handle JSXExpressionContainer
        //     token @ _ => {
        //         unreachable!("Unexpected token: {:?}", token);
        //     }
        // };
        // Box::new(child)
    }

    /**
     * Reads the expression part of a JSX attribute.
     */
    fn read_jsx_attribute_value(&mut self) -> JSXAttributeValue {
        // An explicit assignment. The assigned value can
        // either be a string literal or an identifier in curly brackets
        match self.lexer.next_token() {
            Some(Token::String(literal)) => JSXAttributeValue::StringLiteral(literal),
            Some(Token::LCurlyBracket) => {
                let expr = self.read_expr();
                JSXAttributeValue::Expression(expr)
            }
            token @ _ => unreachable!("Unknown token: {:?}", token),
        }
    }

    /**
     * read_jsx_attr
     *
     * A JSX Attribute comes in two forms:
     *
     * 1. An explicit attribute where the value is passed
     *    in inline (<div id="1" />)
     * 2. An implicit attribute where we assume the attribute
     *    name is a reference to some identifier
     *    (<div id>)
     */
    fn read_jsx_attribute(&mut self) -> JSXAttribute {
        let name = self.read_ident();

        let foo = match self.peek() {
            Some(&Token::Assign) => {
                self.skip_assert(Token::Assign);
                let attribute_value = self.read_jsx_attribute_value();
                JSXAttribute::new(name, attribute_value)
            }
            Some(&Token::Ident(_)) | Some(&Token::ForwardSlash) | Some(&Token::GreaterThan) => {
                let attr_ident = Identifier::new(name.clone());
                let attr_ident_expr = Expression::Identifer(attr_ident);
                let attr_value = JSXAttributeValue::Expression(attr_ident_expr);
                JSXAttribute::new(name, attr_value)
            }
            _ => {
                unreachable!();
            }
        };
        foo
        // println!("read_jsx_attribute:name {:?}", name);
        // match self.lexer.next_token() {
        //     // Explicit assignment
        //     Some(Token::Assign) => {
        //         let attr_value = self.read_jsx_attribute_value();
        //         JSXAttribute::new(name, attr_value)
        //     }
        //     // Shorthand form
        //     Some(Token::Ident(_)) | Some(Token::ForwardSlash) | Some(Token::GreaterThan) => {
        //         println!("shorthand form, creating the identifier and stuff");
        //         let attr_ident = Identifier::new(name.clone());
        //         let attr_ident_expr = Expression::Identifer(attr_ident);
        //         let attr_value = JSXAttributeValue::Expression(attr_ident_expr);
        //         JSXAttribute::new(name, attr_value)
        //     }
        //     token @ _ => unreachable!("Unknown character: {:?}", token),
        // }
    }

    fn read_type_def(&mut self) -> Node {
        // Next token is the identifier for the type definition.
        let name = self.read_ident();
        let def = TypeDefinition::new(name);
        // Just read everything until we reach the end for now...
        loop {
            if self.lexer.next_token() == Some(Token::RCurlyBracket) {
                break;
            }
        }
        Node::TypeDefinition(def)
    }

    /**
     * Inside a module, top level nodes are currently restricted to component
     * and function definitions. This means you cannot have other expressions
     * that are executed when the module is evalulated. All components and
     * functions must be lazily evaluated.
     */
    pub fn next_node(&mut self) -> Node {
        let token = self.lexer.next_token().unwrap();
        match token {
            Token::Reserved(ReservedWord::Component) => self.read_component_declaration(),
            Token::Reserved(ReservedWord::Type) => self.read_type_def(),
            _ => Node::Text,
        }
    }
}
