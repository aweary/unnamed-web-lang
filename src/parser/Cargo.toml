[package]
name = "parser"
version = "0.1.0"
authors = ["Brandon Dail <brandondail@fb.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lexer = { path = "../lexer" }
syntax = { path = "../syntax" }
diagnostics = { path = "../diagnostics" }