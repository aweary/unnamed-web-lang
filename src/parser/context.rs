use crate::ast::{Expr, ExprId, ExprKind, Template, TemplateId};
use crate::ir::{Template as TemplateIR, TemplateInstrList, StaticTemplateId};
use crate::symbol::{Symbol, SymbolTable, SymbolTableEntry, SYMBOL_DEBUG_TABLE_DEV_ONLY};
use crate::typecheck::{infer, TKind};
use id_arena::{Arena};
use string_interner::StringInterner;

use std::cell::RefCell;
use std::collections::{HashMap};

// Trait that allows structs to
pub trait ExprContext {
    fn resolve_expr(&self, id: ExprId) -> &Expr;
}

#[derive(Debug)]
struct StaticTemplateMap {
    cache: HashMap<TemplateInstrList, StaticTemplateId>,
    id: u32,
}

impl StaticTemplateMap {
    pub fn new() -> Self {
        Self {
            cache: HashMap::new(),
            id: 0,
        }
    }

    pub fn insert(&mut self, instrs: TemplateInstrList) -> StaticTemplateId {
        if self.cache.contains_key(&instrs) {
            *self.cache.get(&instrs).unwrap()
        } else {
            let id = StaticTemplateId(self.id);
            self.id += 1;
            self.cache.insert(instrs, id);
            id
        }
    }
}

// "Global" context used for a single parsing session
#[derive(Debug)]
pub struct ParsingContext {
    pub interner: StringInterner<Symbol>,
    // TODO make this private
    symbols: RefCell<SymbolTable>,
    expr_arena: Arena<Expr>,
    template_arena: Arena<Template>,
    type_cache: HashMap<ExprId, TKind>,
    static_template_map: RefCell<StaticTemplateMap>,
    template_ir_map: RefCell<HashMap<TemplateId, TemplateIR>>,
}

impl ExprContext for ParsingContext {
    fn resolve_expr(&self, id: ExprId) -> &Expr {
        self.expr_arena.get(id).unwrap()
    }
}

impl ParsingContext {
    pub fn new() -> Self {
        Self {
            interner: StringInterner::new(),
            expr_arena: Arena::with_capacity(64),
            template_arena: Arena::with_capacity(64),
            symbols: RefCell::new(SymbolTable::new()),
            type_cache: HashMap::new(),
            static_template_map: RefCell::new(StaticTemplateMap::new()),
            template_ir_map: RefCell::new(HashMap::new()),
        }
    }

    pub fn alloc_expr(&mut self, expr: Expr) -> ExprId {
        let id = self.expr_arena.alloc(expr);
        id
    }

    pub fn alloc_template(&mut self, template: Template) -> TemplateId {
        self.template_arena.alloc(template)
    }

    pub fn resolve_expr(&self, id: ExprId) -> &Expr {
        self.expr_arena.get(id).unwrap()
    }

    pub fn resolve_template(&self, id: TemplateId) -> &Template {
        self.template_arena.get(id).unwrap()
    }

    pub fn set_template_ir(&self, id: TemplateId, ir: TemplateIR) {
        self.template_ir_map.borrow_mut().insert(id, ir);
    }

    pub fn alloc_static_template_instrs(&self, instrs: TemplateInstrList) -> StaticTemplateId {
        self.static_template_map.borrow_mut().insert(instrs)
    }

    pub fn has_ty_for(&self, id: ExprId) -> bool {
        self.type_cache.contains_key(&id)
    }

    pub fn begin_scope(&self) {
        self.symbols.borrow_mut().begin_scope();
    }

    pub fn end_scope(&self) {
        self.symbols.borrow_mut().end_scope();
    }

    pub fn symbol_str(&self, sym: Symbol) -> Option<&str> {
        self.interner.resolve(sym)
    }

    pub fn symbol(&mut self, s: &str) -> Symbol {
        // This should only happen in DEV builds
        SYMBOL_DEBUG_TABLE_DEV_ONLY.with(|symbols| {
            let mut interner = symbols.interner.borrow_mut();
            interner.get_or_intern(s);
        });
        self.interner.get_or_intern(s)
    }

    pub fn resolve_type_of(&self, sym: Symbol) -> TKind {
        match self.symbols.borrow().resolve(&sym).unwrap() {
            SymbolTableEntry::Abstract(ty) => *ty,
            SymbolTableEntry::Local(expr_id) => infer(self, *expr_id),
            _ => TKind::Error,
        }
        // Symbol table can have different kinds of entries.
        // Locals resolve to expressions which we can infer the
        // type of. Component/funciton arguments resolve directly
        // to types
    }

    // Returns a static value for an expression if it can be computed
    // at compile time.
    pub fn resolve_value_of(&self, expr_id: ExprId) -> Option<Symbol> {
        let expr = self.resolve_expr(expr_id);
        match *expr.kind {
            ExprKind::Str(text) => Some(text),
            _ => None,
        }
    }

    // pub fn resolve_symbol(&self, sym: Symbol) -> ExprId {
    //     println!("resolving symbol {:?}", sym);
    //     match self.symbols.borrow().resolve(&sym) {
    //         Some(expr_id) => expr_id,
    //         None => {
    //         }
    //     }
    // }

    pub fn define_local(&self, sym: Symbol, expr_id: ExprId) {
        self.symbols.borrow_mut().define_local(sym, expr_id);
    }

    pub fn define_abstract(&self, sym: Symbol, ty: TKind) {
        self.symbols.borrow_mut().define_abstract(sym, ty);
    }

    pub fn is_defined(&self, sym: &Symbol) -> bool {
        self.symbols.borrow().resolve(sym).is_some()
    }

    pub fn type_of(&self, expr_id: ExprId) -> TKind {
        match self.type_cache.get(&expr_id) {
            Some(ty) => return *ty,
            _ => (),
        };
        infer(self, expr_id)
    }

    pub fn report_err(&self, message: &str) {
        println!("report error: {}", message)
    }
}
