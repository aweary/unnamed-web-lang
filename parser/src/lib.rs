#![warn(clippy::all, clippy::pedantic)]
#![allow(clippy::must_use_candidate)]

mod parser;

pub use crate::parser::*;

#[cfg(test)]
mod tests {
    use crate::parser::Parser;
    use insta::assert_json_snapshot;
    use syntax::ast;

    fn parse_module(source: &str) -> ast::Module {
        let mut parser = Parser::new(source);
        parser.parse_module().unwrap()
    }

    #[test]
    fn fixture_function_simple() {
        let input = include_str!("./inputs/function_simple.dan");
        assert_json_snapshot!(parse_module(input));
    }

    #[test]
    fn fixture_control_flow() {
        let input = include_str!("./inputs/control_flow.dan");
        assert_json_snapshot!(parse_module(input));
    }

    #[test]
    fn fixture_trailing_closure_control_flow() {
        let input = include_str!("./inputs/trailing_closures.dan");
        assert_json_snapshot!(parse_module(input));
    }
}
