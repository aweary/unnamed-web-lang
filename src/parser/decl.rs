use super::stmt::StmtParser;

use crate::ast::{Decl, DeclKind, FuncDecl, Param};
use crate::error::ParseError;
use crate::parser::Parser;
use crate::result::Result;
use crate::typecheck::TKind;

use crate::token::Keyword::Func;
use crate::token::TokenKind::{
    Colon, Comma, GreaterThan, Ident, Keyword, LParen, LessThan, RParen, Semi,
};

/**
 * Declarations are technically statements, but we currently
 * restrict where you can make certain kinds of declarations,
 * so its easier to separate those out.
 */
pub trait DeclParser<'a>: StmtParser<'a> {
    fn decl(&mut self) -> Result<Decl>;
    fn decl_list(&mut self) -> Result<Vec<Decl>>;
    fn fn_decl(&mut self) -> Result<FuncDecl>;
    fn fn_param(&mut self) -> Result<Param>;
    fn fn_param_list(&mut self) -> Result<Option<Vec<Param>>>;
}

impl<'a> DeclParser<'a> for Parser<'a> {
    fn decl_list(&mut self) -> Result<Vec<Decl>> {
        let mut decls = vec![];
        while !self.peek_token()?.follow_stmt_list() {
            decls.push(self.decl()?);
            self.eat(Semi)?;
        }
        Ok(decls)
    }

    fn decl(&mut self) -> Result<Decl> {
        match self.peek_token()?.kind {
            Keyword(Func) => Ok(Decl {
                kind: DeclKind::Func(self.fn_decl()?),
            }),
            _ => return Err(ParseError::Unimplemented),
        }
    }

    fn fn_param(&mut self) -> Result<Param> {
        let name = self.ident()?;
        self.expect(Colon)?;
        let ident = self.type_ident()?;
        // This maps the tokens representing the built-in types
        // of the environment, to the internal representation
        // the type system uses (TKind).
        // If a new type is added, it needs to go here to be parsed.
        let ty = match self.ctx.symbol_str(ident) {
            Some("int") | Some("number") => TKind::Number,
            s => {
                println!("SSS {:?}",s);
                TKind::Error
            },
        };
        // let ty = TKind::Number;
        Ok(Param { name, ty })
    }

    fn fn_param_list(&mut self) -> Result<Option<Vec<Param>>> {
        let mut params = vec![];
        loop {
            match self.peek_token()?.kind {
                RParen => {
                    self.expect(RParen)?;
                    break;
                }
                Ident(_) => {
                    let param = self.fn_param()?;
                    params.push(param);
                    if !self.eat(Comma)? {
                        self.expect(RParen)?;
                        break;
                    }
                }
                _ => return Err(ParseError::UnexpectedToken(self.next_token().unwrap())),
            }
        }
        Ok(if params.is_empty() {
            None
        } else {
            Some(params)
        })
    }

    fn fn_decl(&mut self) -> Result<FuncDecl> {
        self.expect(Keyword(Func))?;
        let name = self.ident()?;
        // Optional type arguments
        if self.eat(LessThan)? {
            // One type argument for now
            let ty = self.type_ident()?;
            println!("ty: {:?}", ty);
            self.expect(GreaterThan)?;
        }
        self.expect(LParen)?;
        let params = self.fn_param_list()?;
        self.expect(Colon)?;
        let return_ty = self.type_ident()?;
        let block = self.block()?;
        Ok(FuncDecl {
            name,
            params,
            return_ty,
            block,
        })
    }
}
