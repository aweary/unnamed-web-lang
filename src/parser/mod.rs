mod decl;
mod expr;
mod stmt;
mod context;

use decl::DeclParser;
// use expr::ExprParser;
// use stmt::StmtParser;
pub use context::{ParsingContext, ExprContext};


use crate::error::ParseError;
use crate::lexer::Lexer;
use crate::result::Result;
use crate::ast::{Module, Program};
use crate::symbol::Symbol;
use crate::token::{Token, TokenKind};
use std::rc::Rc;
use std::cell::RefCell;

// temporary use so that the Rust compiler checks this module
// use crate::ir::*;

pub fn parse_program<'a>(source: &'a str, mut ctx: &'a mut ParsingContext) -> Result<Program> {
    let mut parser = Parser::new(&source, &mut ctx);
    parser.parse()
}

pub struct Parser<'a> {
    lexer: Lexer<'a>,
    ctx: &'a mut ParsingContext,
}

impl<'a> Parser<'a> {
    pub fn new(source: &'a str, ctx: &'a mut ParsingContext) -> Self {
        let lexer = Lexer::new(&source);
        Parser { lexer, ctx }
    }

    fn next_token(&mut self) -> Result<Token> {
        match self.lexer.next_token(&mut self.ctx) {
            Ok(token) => {
                Ok(token)
            }
            Err(_) => Err(ParseError::LexError),
        }
    }

    fn peek_token(&mut self) -> Result<&Token> {
        match self.lexer.peek_token(&mut self.ctx) {
            Ok(token) => Ok(token),
            Err(_) => Err(ParseError::LexError),
        }
    }

    fn eat(&mut self, kind: TokenKind) -> Result<bool> {
        if self.peek_token()?.kind == kind {
            self.expect(kind)?;
            Ok(true)
        } else {
            Ok(false)
        }
    }

    fn expect(&mut self, kind: TokenKind) -> Result<Token> {
        let token = self.next_token()?;
        if token.kind != kind {
            Err(ParseError::UnexpectedToken(token))
        } else {
            Ok(token)
        }
    }

    fn ident(&mut self) -> Result<Symbol> {
        let token = self.next_token()?;
        match token.kind {
            TokenKind::Ident(sym) => Ok(sym),
            _ => Err(ParseError::UnexpectedToken(token)),
        }
    }

    /**
     * This is currently an alias to `ident`, but a type
     * identifier has different syntax. For example, polymorphic
     * types have type parameters.
     */
    fn type_ident(&mut self) -> Result<Symbol> {
        self.ident()
    }

    pub fn parse(&mut self) -> Result<Program> {
        // top-level statements
        let stmts = self.decl_list()?;
        let module = Module { stmts };
        let modules = vec![module];
        let program = Program { modules };
        Ok(program)
    }
}
