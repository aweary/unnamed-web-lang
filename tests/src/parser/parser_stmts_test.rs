#[cfg(test)]
mod tests {
    use parser::Parser;
    use std::fs;
    use insta::{glob, assert_json_snapshot};
    #[test]
    // Easy but very noisy testing. Hard to debug when something fails.
    // Good for now, but more granular tests in the future will be great.
    fn parser_tests() {
        glob!("fixtures/*.dan", |path| {
            let input = fs::read_to_string(path).unwrap();
            let mut parser = Parser::new(&input);
            let ast = parser.parse_module().unwrap();
            assert_json_snapshot!(ast);
        });
    }
}