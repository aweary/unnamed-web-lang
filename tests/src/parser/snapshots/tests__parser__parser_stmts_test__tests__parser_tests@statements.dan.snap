---
source: tests/src/parser/parser_stmts_test.rs
expression: ast
input_file: tests/src/parser/fixtures/statements.dan
---
{
  "items": [
    {
      "kind": {
        "Fn": {
          "name": {
            "symbol": {
              "unstable_source": "main"
            },
            "span": {
              "start": 3,
              "end": 7
            }
          },
          "params": "Empty",
          "body": {
            "stmts": [
              {
                "id": 0,
                "kind": {
                  "Local": {
                    "id": 0,
                    "name": {
                      "Ident": [
                        {
                          "symbol": {
                            "unstable_source": "a"
                          },
                          "span": {
                            "start": 20,
                            "end": 21
                          }
                        },
                        {
                          "start": 20,
                          "end": 21
                        }
                      ]
                    },
                    "ty": null,
                    "init": {
                      "kind": {
                        "Lit": {
                          "span": {
                            "start": 24,
                            "end": 25
                          },
                          "kind": {
                            "Number": {
                              "unstable_source": "1"
                            }
                          }
                        }
                      },
                      "span": {
                        "start": 24,
                        "end": 25
                      }
                    },
                    "span": {
                      "start": 16,
                      "end": 25
                    }
                  }
                },
                "span": {
                  "start": 16,
                  "end": 26
                },
                "has_semi": true
              },
              {
                "id": 0,
                "kind": {
                  "If": {
                    "span": {
                      "start": 31,
                      "end": 97
                    },
                    "condition": {
                      "kind": {
                        "Lit": {
                          "span": {
                            "start": 34,
                            "end": 38
                          },
                          "kind": {
                            "Bool": {
                              "unstable_source": "true"
                            }
                          }
                        }
                      },
                      "span": {
                        "start": 34,
                        "end": 38
                      }
                    },
                    "block": {
                      "stmts": [
                        {
                          "id": 0,
                          "kind": {
                            "Local": {
                              "id": 0,
                              "name": {
                                "Ident": [
                                  {
                                    "symbol": {
                                      "unstable_source": "b"
                                    },
                                    "span": {
                                      "start": 53,
                                      "end": 54
                                    }
                                  },
                                  {
                                    "start": 53,
                                    "end": 54
                                  }
                                ]
                              },
                              "ty": null,
                              "init": {
                                "kind": {
                                  "Lit": {
                                    "span": {
                                      "start": 57,
                                      "end": 58
                                    },
                                    "kind": {
                                      "Number": {
                                        "unstable_source": "2"
                                      }
                                    }
                                  }
                                },
                                "span": {
                                  "start": 57,
                                  "end": 58
                                }
                              },
                              "span": {
                                "start": 49,
                                "end": 58
                              }
                            }
                          },
                          "span": {
                            "start": 49,
                            "end": 59
                          },
                          "has_semi": true
                        }
                      ],
                      "id": 0,
                      "span": {
                        "start": 39,
                        "end": 65
                      }
                    },
                    "alt": {
                      "Block": {
                        "stmts": [
                          {
                            "id": 0,
                            "kind": {
                              "Local": {
                                "id": 0,
                                "name": {
                                  "Ident": [
                                    {
                                      "symbol": {
                                        "unstable_source": "c"
                                      },
                                      "span": {
                                        "start": 85,
                                        "end": 86
                                      }
                                    },
                                    {
                                      "start": 85,
                                      "end": 86
                                    }
                                  ]
                                },
                                "ty": null,
                                "init": {
                                  "kind": {
                                    "Lit": {
                                      "span": {
                                        "start": 89,
                                        "end": 90
                                      },
                                      "kind": {
                                        "Number": {
                                          "unstable_source": "3"
                                        }
                                      }
                                    }
                                  },
                                  "span": {
                                    "start": 89,
                                    "end": 90
                                  }
                                },
                                "span": {
                                  "start": 81,
                                  "end": 90
                                }
                              }
                            },
                            "span": {
                              "start": 81,
                              "end": 91
                            },
                            "has_semi": true
                          }
                        ],
                        "id": 0,
                        "span": {
                          "start": 71,
                          "end": 97
                        }
                      }
                    }
                  }
                },
                "span": {
                  "start": 31,
                  "end": 97
                },
                "has_semi": false
              }
            ],
            "id": 0,
            "span": {
              "start": 10,
              "end": 99
            }
          },
          "return_ty": null,
          "is_async": false,
          "generics": null,
          "span": {
            "start": 0,
            "end": 99
          }
        }
      },
      "span": {
        "start": 0,
        "end": 99
      }
    }
  ]
}
